File: Basics/Trie Implementation/trie.go
package main

import "fmt"

type TrieNode struct {
	children       map[rune]*TrieNode
	isEndOfWord    bool
}

func NewTrieNode() *TrieNode {
	return &TrieNode{
		children:    make(map[rune]*TrieNode),
		isEndOfWord: false,
	}
}

type Trie struct {
	root *TrieNode
}

func NewTrie() *Trie {
	return &Trie{root: NewTrieNode()}
}

func (t *Trie) Insert(word string) {
	node := t.root
	for _, char := range word {
		if _, exists := node.children[char]; !exists {
			node.children[char] = NewTrieNode()
		}
		node = node.children[char]
	}
	node.isEndOfWord = true
}

func (t *Trie) Search(word string) bool {
	node := t.root
	for _, char := range word {
		if _, exists := node.children[char]; !exists {
			return false
		}
		node = node.children[char]
	}
	return node.isEndOfWord
}

func (t *Trie) StartsWith(prefix string) bool {
	node := t.root
	for _, char := range prefix {
		if _, exists := node.children[char]; !exists {
			return false
		}
		node = node.children[char]
	}
	return true
}

// Example usage:
func main() {
	trie := NewTrie()
	trie.Insert("apple")
	fmt.Println(trie.Search("apple"))     // Output: true
	fmt.Println(trie.Search("app"))       // Output: false
	fmt.Println(trie.StartsWith("app"))   // Output: true
	trie.Insert("app")
	fmt.Println(trie.Search("app"))       // Output: true
}

File: Basics/Trie Implementation/trie.rs
use std::collections::HashMap;

struct TrieNode {
    children: HashMap<char, TrieNode>,
    is_end_of_word: bool,
}

impl TrieNode {
    fn new() -> Self {
        TrieNode {
            children: HashMap::new(),
            is_end_of_word: false,
        }
    }
}

struct Trie {
    root: TrieNode,
}

impl Trie {
    fn new() -> Self {
        Trie { root: TrieNode::new() }
    }

    fn insert(&mut self, word: &str) {
        let mut node = &mut self.root;
        for char in word.chars() {
            node = node.children.entry(char).or_insert(TrieNode::new());
        }
        node.is_end_of_word = true;
    }

    fn search(&self, word: &str) -> bool {
        let mut node = &self.root;
        for char in word.chars() {
            if let Some(next_node) = node.children.get(&char) {
                node = next_node;
            } else {
                return false;
            }
        }
        node.is_end_of_word
    }

    fn starts_with(&self, prefix: &str) -> bool {
        let mut node = &self.root;
        for char in prefix.chars() {
            if let Some(next_node) = node.children.get(&char) {
                node = next_node;
            } else {
                return false;
            }
        }
        true
    }
}

// Example usage:
fn main() {
    let mut trie = Trie::new();
    trie.insert("apple");
    println!("{}", trie.search("apple"));  // Output: true
    println!("{}", trie.search("app"));    // Output: false
    println!("{}", trie.starts_with("app"));  // Output: true
    trie.insert("app");
    println!("{}", trie.search("app"));    // Output: true
}
